tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc

```diff
@@ -5805,6 +5817,553 @@ Status ConvertBatchMatMul(OpConverterParams* params) {
+  while (begin != end) {
+    nvinfer1::IElementWiseLayer* layer =
+        params->converter->network()->addElementWise(
+            **out, **begin, nvinfer1::ElementWiseOperation::kPROD);
+    TFTRT_RETURN_ERROR_IF_NULLPTR(layer, params->node_def.name());
+    *out = layer->getOutput(0);
+    begin++;
+  }
+  return Status::OK();
+}
+
+Status ConcatenateShape(OpConverterParams* params,
+                        const std::vector<nvinfer1::ITensor*> size_tensors,
+                        nvinfer1::ITensor** new_shape) {
+  nvinfer1::IConcatenationLayer* layer =
+      params->converter->network()->addConcatenation(
+          const_cast<nvinfer1::ITensor* const*>(size_tensors.data()),
+          size_tensors.size());
+  TFTRT_RETURN_ERROR_IF_NULLPTR(layer, params->node_def.name());
+  layer->setAxis(0);
+  *new_shape = layer->getOutput(0);
+  return Status::OK();
+}
+
+// Reshape operand so that the free dimensions are combined into a single dim,
+// and the contract dimensions are combined into another single dim.
+Status GetEinsumNewDynamicShape(OpConverterParams* params,
+                                const EinsumDescriptor& desc,
+                                nvinfer1::ITensor** new_shape) {
+  std::vector<nvinfer1::ITensor*> size(desc.size_tensors.begin(),
+                                       desc.size_tensors.begin() + desc.n + 2);
+
+  int idx_f = desc.layout == EinsumLayout::BFC ? desc.n : desc.n + 1;
+  int idx_c = desc.layout == EinsumLayout::BFC ? desc.n + 1 : desc.n;
+
+  TF_RETURN_IF_ERROR(GetDimsProdDynamic(
+      params, desc.size_tensors.begin() + desc.offset_f,
+      desc.size_tensors.begin() + desc.offset_f + desc.f, &size[idx_f]));
+
+  TF_RETURN_IF_ERROR(GetDimsProdDynamic(
+      params, desc.size_tensors.begin() + desc.offset_c,
+      desc.size_tensors.begin() + desc.offset_c + desc.c, &size[idx_c]));
+
+  TF_RETURN_IF_ERROR(ConcatenateShape(params, size, new_shape));
+  return Status::OK();
+}
+
+// Reshape operand so that the free dimensions are combined into a single dim,
+// and the contract dimensions are combined into another single dim.
+Status GetEinsumNewStaticShape(const EinsumDescriptor& desc,
+                               nvinfer1::Dims* new_dims) {
+  new_dims->nbDims = desc.n + 2;
+  // Copy batch dims.
+  std::copy(desc.dims.d, desc.dims.d + desc.n, new_dims->d);
+  // Combine free dims and contract dims.
+  int idx_f = desc.layout == EinsumLayout::BFC ? desc.n : desc.n + 1;
+  int idx_c = desc.layout == EinsumLayout::BFC ? desc.n + 1 : desc.n;
+  TF_RETURN_IF_ERROR(
+      GetDimsProd(desc.dims, desc.offset_f, desc.f, new_dims->d + idx_f));
+  TF_RETURN_IF_ERROR(
+      GetDimsProd(desc.dims, desc.offset_c, desc.c, new_dims->d + idx_c));
+  return Status::OK();
+}
+
+// Add shuffle layer (if needed) to bring einsum operand to a matmul compatible
+// format.
+Status ShuffleEinsumTensor(OpConverterParams* params,
+                           std::unique_ptr<TRT_TensorOrWeights>& operand,
```

Comment:
We shouldn't use non-const reference, but pointer instead.

Reply:
Changed.